sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextFactory.java

```diff
@@ -212,13 +212,12 @@ private static <T> Mono<Pair<List<Range<String>>, QueryInfo>> getTargetRangesFro
                 queryInfo));
     }
 
-    private static void tryCacheQueryPlan(
+    synchronized private static void tryCacheQueryPlan(
         SqlQuerySpec query,
         PartitionedQueryExecutionInfo partitionedQueryExecutionInfo,
         Map<String, PartitionedQueryExecutionInfo> queryPlanCache) {
-        QueryInfo queryInfo = partitionedQueryExecutionInfo.getQueryInfo();
-        if (canCacheQuery(queryInfo) && !queryPlanCache.containsKey(query.getQueryText())) {
-            if (queryPlanCache.size() == Constants.QUERYPLAN_CACHE_SIZE) {
+        if (canCacheQuery(partitionedQueryExecutionInfo.getQueryInfo()) && !queryPlanCache.containsKey(query.getQueryText())) {
```

Comment:
I don't quite understand from this change (or the PR description) how this race condition can result in an OutOfMemoryError.
I get the race condition - and this was understood and intentionally accepted - because even if the size check might not reflect concurrent updates to result in an OutOfMemoryError millions of puts would need to escape the check - which isn't realistic. 
Maybe the error details would help to clarify why this change would ever address an OutOfMemoryError?

Reply:
Updated the description, please check!

Reply:
Thanks - LGTM